Modules/_datetimemodule.c

```diff
@@ -637,17 +637,22 @@ check_date_args(int year, int month, int day)
 {
 
     if (year < MINYEAR || year > MAXYEAR) {
-        PyErr_Format(PyExc_ValueError, "year %i is out of range", year);
+        PyErr_Format(PyExc_ValueError,
+                     "year must be in %d..%d, but got %d",
+                     MINYEAR, MAXYEAR, year);
         return -1;
     }
     if (month < 1 || month > 12) {
-        PyErr_SetString(PyExc_ValueError,
-                        "month must be in 1..12");
+        PyErr_Format(PyExc_ValueError,
+                    "month must be in 1..12, but got %d",
+                    month);
         return -1;
     }
-    if (day < 1 || day > days_in_month(year, month)) {
-        PyErr_SetString(PyExc_ValueError,
-                        "day is out of range for month");
+    int dim = days_in_month(year, month);
+    if (day < 1 || day > dim) {
```

Comment:
Observation: previously, `days_in_month` was only called if `day >= 1` (probably as an optimisation).

Reply:
yes, most likely `day < 1` occurs less often than `day >= 1` by more than two times, so an second call to `days_in_month()` to report an error would be a more optimal option

Reply:
Oh, i got you and I have a suggestion to use lazy evaluation, ensuring the function is invoked only if day < 1 or day > dim, something like this:
```c
int dim = 0;
if (day < 1 || (day > (dim = days_in_month(year, month)))) {
    if (dim == 0) {
        dim = days_in_month(year, month);
    }
    PyErr_Format(PyExc_ValueError,
                 "day must be in 1..%d, but got %d", dim, day);
    return -1;
}
```

Reply:
That would work, but I'm not sure it really matters; I doubt anyone would notice it on a benchmark. Moreover, readability counts.

Reply:
Let's just keep it like it is :)